[scroll-animations] Crash under AnimationEffectTiming::getBasicTiming

nmoucht opened the pull request:
#### 71bd0d4b8e9cba4a0f9712401d50a8ff3d7ceefd
<pre>
[scroll-animations] Crash under AnimationEffectTiming::getBasicTiming
<a href="https://bugs.webkit.org/show_bug.cgi?id=292742">https://bugs.webkit.org/show_bug.cgi?id=292742</a>
<a href="https://rdar.apple.com/149951894">rdar://149951894</a>

Reviewed by Antoine Quint.

Under AnimationEffectTiming::getBasicTiming, there is an assertion that if we have a resolved
start time for an animation, the animation&apos;s timeline should have a resolved current time. It
is fairly easy to get into the case where this doesn&apos;t hold true, for a scroll driven animation
where the scroll container becomes non-scrollable after the start time has been resolved. My
original solution was to make the start time unresolved in WebAnimation::autoAlignStartTime
if we saw that the timeline&apos;s time was unresolved, but after <a href="https://commits.webkit.org/294121@main">https://commits.webkit.org/294121@main</a>
changed the timing of when we update the timeline&apos;s cached current time, it is possible for the
cached time to become unresolved before we can run WebAnimation::autoAlignStartTime. A simpler
solution is to remove the assert and check that both the start time and the timeline&apos;s current
time are resolved.

* LayoutTests/TestExpectations:
* LayoutTests/animations/sda-crash-expected.txt: Added.
* LayoutTests/animations/sda-crash.html: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/animation/AnimationEffectTiming.cpp:
(WebCore::AnimationEffectTiming::getBasicTiming const):

Canonical link: <a href="https://commits.webkit.org/294813@main">https://commits.webkit.org/294813@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd2ad7598e0a6df68107b04ffeb8d01ef89de39f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103156 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22831 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108360 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53799 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105195 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23166 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31331 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78443 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35361 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106162 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17960 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93079 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58776 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17806 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11121 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53154 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87613 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11184 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110701 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30293 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22335 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87410 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30658 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89278 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87040 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31899 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9613 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24617 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16740 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30220 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35542 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30028 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33355 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31590 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts reviewed (approved):
Looks good Nikos and thanks for taking care of this. Would it not be possible to create a test that does not rely on `UIHelper.waitForScrollCompletion()` and instead use `scrollTop` to set the scroll position? This is how most scroll-driven animations tests work. This could also be made a WPT.

nmoucht committed:
[scroll-animations] Crash under AnimationEffectTiming::getBasicTiming
https://bugs.webkit.org/show_bug.cgi?id=292742
rdar://149951894

Reviewed by Antoine Quint.

Under AnimationEffectTiming::getBasicTiming, there is an assertion that if we have a resolved
start time for an animation, the animation's timeline should have a resolved current time. It
is fairly easy to get into the case where this doesn't hold true, for a scroll driven animation
where the scroll container becomes non-scrollable after the start time has been resolved. My
original solution was to make the start time unresolved in WebAnimation::autoAlignStartTime
if we saw that the timeline's time was unresolved, but after https://commits.webkit.org/294121@main
changed the timing of when we update the timeline's cached current time, it is possible for the
cached time to become unresolved before we can run WebAnimation::autoAlignStartTime. A simpler
solution is to remove the assert and check that both the start time and the timeline's current
time are resolved.

* LayoutTests/TestExpectations:
* LayoutTests/animations/sda-crash-expected.txt: Added.
* LayoutTests/animations/sda-crash.html: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/animation/AnimationEffectTiming.cpp:
(WebCore::AnimationEffectTiming::getBasicTiming const):

Canonical link: https://commits.webkit.org/294813@main
